packages/core/source/legacy/host.ts

```diff
@@ -0,0 +1,335 @@
+import {
+  KIND_TEXT as LEGACY_KIND_TEXT,
+  KIND_FRAGMENT as LEGACY_KIND_FRAGMENT,
+  ACTION_MOUNT as LEGACY_ACTION_MOUNT,
+  ACTION_INSERT_CHILD as LEGACY_ACTION_INSERT_CHILD,
+  ACTION_REMOVE_CHILD as LEGACY_ACTION_REMOVE_CHILD,
+  ACTION_UPDATE_PROPS as LEGACY_ACTION_UPDATE_PROPS,
+  ACTION_UPDATE_TEXT as LEGACY_ACTION_UPDATE_TEXT,
+  type RemoteChannel as LegacyRemoteChannel,
+  type ActionArgumentMap as LegacyActionArgumentMap,
+  type RemoteComponentSerialization as LegacyRemoteComponentSerialization,
+  type RemoteTextSerialization as LegacyRemoteTextSerialization,
+  type RemoteFragmentSerialization as LegacyRemoteFragmentSerialization,
+} from '@remote-ui/core';
+
+import type {
+  RemoteMutationRecord,
+  RemoteTextSerialization,
+  RemoteElementSerialization,
+  RemoteConnection,
+  RemoteNodeSerialization,
+} from '../types';
+import {
+  ROOT_ID,
+  NODE_TYPE_TEXT,
+  NODE_TYPE_ELEMENT,
+  MUTATION_TYPE_INSERT_CHILD,
+  MUTATION_TYPE_REMOVE_CHILD,
+  MUTATION_TYPE_UPDATE_PROPERTY,
+  MUTATION_TYPE_UPDATE_TEXT,
+} from '../constants.ts';
+
+export interface LegacyRemoteChannelElementMap {
+  [key: string]: string;
+}
+
+export interface LegacyRemoteChannelOptions {
+  elements?: LegacyRemoteChannelElementMap;
+}
+
+/**
+ * Adapts a Remote DOM `RemoteConnection` object into a `remote-ui` `RemoteChannel`.
+ * This allows you to use a Remote DOM receiver class on the host, even if the remote
+ * environment is using `remote-ui`.
+ *
+ * @example
+ * ```tsx
+ * import {DOMRemoteReceiver} from '@remote-dom/core/receivers';
+ * import {adaptToLegacyRemoteChannel} from '@remote-dom/core/legacy';
+ *
+ * const receiver = new DOMRemoteReceiver();
+ * const channel = adaptToLegacyRemoteChannel(receiver.connection);
+ *
+ * // Do something with the channel
+ * sendChannelToRemoteEnvironment(channel);
+ * ```
+ */
+export function adaptToLegacyRemoteChannel(
+  connection: RemoteConnection,
+  options?: LegacyRemoteChannelOptions,
+): LegacyRemoteChannel {
+  const tree: Record<string, Array<{id: string; slot?: string}>> = {};
+
+  function persistNode(
```

Comment:
Nit: can you move the returned function first, and have these functions appear after it? They all get hoisted up, but the reversed order puts the more important thing first, so you can see how these are used before seeing their implementation.